src/backend/distributed/utils/background_jobs.c

```diff
@@ -0,0 +1,1238 @@
+ * environment to the executor.
+ */
+static dsm_segment *
+StoreArgumentsInDSM(char *database, char *username, char *command, int64 taskId)
+{
+	/*
+	 * Create the shared memory that we will pass to the background
+	 * worker process.  We use DSM_CREATE_NULL_IF_MAXSEGMENTS so that we
+	 * do not ERROR here.  This way, we can mark the job as failed and
+	 * keep the launcher process running normally.
+	 */
+	shm_toc_estimator e = { 0 };
+	shm_toc_initialize_estimator(&e);
+	shm_toc_estimate_chunk(&e, strlen(database) + 1);
+	shm_toc_estimate_chunk(&e, strlen(username) + 1);
+	shm_toc_estimate_chunk(&e, strlen(command) + 1);
+#define QUEUE_SIZE ((Size) 65536)
+	shm_toc_estimate_chunk(&e, QUEUE_SIZE);
+	shm_toc_estimate_chunk(&e, sizeof(int64));
+	shm_toc_estimate_keys(&e, CITUS_BACKGROUND_TASK_NKEYS);
+	Size segsize = shm_toc_estimate(&e);
+
+	dsm_segment *seg = dsm_create(segsize, DSM_CREATE_NULL_IF_MAXSEGMENTS);
+	if (seg == NULL)
+	{
+		ereport(ERROR,
+				(errmsg("max number of DSM segments may has been reached")));
+
+		return NULL;
+	}
+
+	shm_toc *toc = shm_toc_create(CITUS_BACKGROUND_TASK_MAGIC, dsm_segment_address(seg),
+								  segsize);
+
+	Size size = strlen(database) + 1;
+	char *databaseTarget = shm_toc_allocate(toc, size);
+	strcpy_s(databaseTarget, size, database);
+	shm_toc_insert(toc, CITUS_BACKGROUND_TASK_KEY_DATABASE, databaseTarget);
+
+	size = strlen(username) + 1;
+	char *usernameTarget = shm_toc_allocate(toc, size);
+	strcpy_s(usernameTarget, size, username);
+	shm_toc_insert(toc, CITUS_BACKGROUND_TASK_KEY_USERNAME, usernameTarget);
+
+	size = strlen(command) + 1;
+	char *commandTarget = shm_toc_allocate(toc, size);
+	strcpy_s(commandTarget, size, command);
+	shm_toc_insert(toc, CITUS_BACKGROUND_TASK_KEY_COMMAND, commandTarget);
+
+	shm_mq *mq = shm_mq_create(shm_toc_allocate(toc, QUEUE_SIZE), QUEUE_SIZE);
+	shm_toc_insert(toc, CITUS_BACKGROUND_TASK_KEY_QUEUE, mq);
+	shm_mq_set_receiver(mq, MyProc);
+
+	int64 *taskIdTarget = shm_toc_allocate(toc, sizeof(int64));
+	*taskIdTarget = taskId;
+	shm_toc_insert(toc, CITUS_BACKGROUND_TASK_KEY_TASK_ID, taskIdTarget);
+
+	/*
+	 * Attach the queue before launching a worker, so that we'll automatically
+	 * detach the queue if we error out.  (Otherwise, the worker might sit
+	 * there trying to write the queue long after we've gone away.)
+	 */
+	MemoryContext oldcontext = MemoryContextSwitchTo(TopMemoryContext);
+	shm_mq_attach(mq, seg, NULL);
+	MemoryContextSwitchTo(oldcontext);
+
+	return seg;
+}
+
+
+/*
+ * StartCitusBackgroundJobExecuter start a new background worker for the execution of a
+ * background task. Users intereted in the shared memory segment that is created between
+ * the background worker and the current backend can pass in a segOut to get a pointer to
+ * the dynamic shared memory.
+ */
+static BackgroundWorkerHandle *
+StartCitusBackgroundJobExecuter(char *database, char *user, char *command, int64 taskId,
```

Comment:
This thing is executing a task not a job.

```suggestion
StartCitusBackgroundTaskExecuter(char *database, char *user, char *command, int64 taskId,
```

same for all the related functions